app/models/result.rb

```diff
@@ -59,7 +59,7 @@ def get_subtotal(user_visibility = :ta)
     unless marks.empty?
       assignment = submission.grouping.assignment
       assignment.get_criteria(user_visibility).each do |criterion|
-        mark = marks.find_by(markable_id: criterion.id)
+        mark = marks.find_by(markable_id: criterion.id, markable_type: criterion.class.to_s)
```

Comment:
This brings to light other issues. Since it seems possible to have different types of criteria with the same id, we should create marks with the appropriate markable_type, which may require changes in other areas in this file such as in the callback method create_marks, and in check_for_nil_marks. Other files that deal with the creation or checking of marks may also need to be modified so that they are not only associated to a particular markable_id, but also to a markable_type. Line 514 in the results_controller for example.

Also I looked at the test that started failing with this change and it fails because the mark that we get now has a different markable_type, so I guess we have to double check both the tests and the code. Unless it is feasible to make sure we have different ids across criteria.

I hope this helps.